### PR TITLE
Enable distributed strategies in tests.

### LINKED ISSUE #1682: Enable multiprocessing when testing with GPU and support distributed strategies in the tests.
**Describe the feature and the current behavior/state.**

Here I'm not going to discuss the bazel case as it's much more complicated to handle, and we currently advertize using pytest anyway to run the tests. We can of course make sure everything is compatible though.

This revamping of gpu testing has multiple objectives:

* The tests should behave the same weither the contributor has a gpu or not. Meaning we shouldn't run all the tests on a gpu just because a gpu is available, otherwise it hurts reproducibility.
* The test suite should be able to run with multiple workers in kokoro or when a user has multiple gpus. Pytest should use all gpus visible by the main process.
* We need to support testing with distributed strategies. Currently it doesn't work. A fix has been started in #1209 but we need to update it for pytest. 
* Making the whole thing simple to use and to maintain. Notably, we would get rid of this file: https://github.com/tensorflow/addons/blob/master/tools/testing/parallel_gpu_execute.sh which is quite hard to work on.

To do all that, here is my proposal:

Stuff to know: 
* Pytest-xdist uses multiprocessing to run the tests, not multithreading. 
* 2 environement variables are available in each of the workers to identify them. https://github.com/pytest-dev/pytest-xdist#identifying-the-worker-process-during-a-test

### Test workers
Suppose we have a machine with 10CPUs and 4 GPUs, 10 processes will start to run the test suite. Workers 0 to 3 will have ownership of one GPU each (we can use CUDA_VISIBLE_DEVICES to enforce that, but I'm not even sure that's needed with the proposed implementation). Workers 4 to 9 will have no gpu available.

### Virtual devices
Each of those processes, when starting, will split their physical device into 2 virtual device. Tests that just need to run on gpu will use the first of those virtual devices. Processes which need to test distributed strategies will use the two of them. We assume here that 2 virtual devices are enough to test distributed strategies.

### Impact on the contributors:
For this whole machinery to work, we need to know which test needs to run on CPU, GPU, or in distributed strategies. To do that we'll use pytest markers: `@pytest.mark.....`

* By default, if no marker is found, the test will run on CPU: `with device("CPU:0")`. It's equivalent to 
`@pytest.mark.run_on(["cpu"])`.
* To run with gpu only: `@pytest.mark.run_on(["gpu"])`.
* To run on the cpu and gpu: `@pytest.mark.run_on(["cpu", "gpu"])` (test runs twice)
* To run in within a distributed strategy `@pytest.mark.run_on(["distributed strategy"])`. (runs once here).
* To run with everything `@pytest.mark.run_on(["cpu", "gpu", "distributed strategy"])`
* To make crazy stuff, and not run the test in any device scope: `@pytest.mark.no_device_scope`. Then the contributor can do whatever he/she wants in the test.

Of course, if no gpu is available, we just skip the tests needing a distribution strategy or the gpu. Contributors who handle the devices manually have to make sure to skip manually the test if the gpu is used.

Since gpu are often the scarsest ressource (nb gpus << nb cpus), tests needing the gpu will also be marked with `@pytest.mark.tryfirst` to ensure that we don't have workers starvation at the end (to get maximum speed).

To implement that, we need first to convert all tests to pytest (as opposed to unittest) it's currently 80% done and thanks a lot @autoih for putting a LOT of work into that.

**Relevant information**
- Are you willing to contribute it (yes/no): yes
- Are you willing to maintain it going forward? (yes/no): yes
- Is there a relevant academic paper? (if so, where): no
- Is there already an implementation in another framework? (if so, where): no
- Was it part of tf.contrib? (if so, where): no

**Which API type would this fall under (layer, metric, optimizer, etc.)**

Testing

**Who will benefit with this feature?**

Contributors with gpu, CI.

**Any other info.**

I believe that the implementation will first go in tensorflow addons because we have 4 GPUs available in the CI. Later on when it's stable we can split it from tensorflow addons and make it a separate pytest plugin with a public API.

Comments welcome. Especially from @Squadrick , @hyang0129 , @seanpmorgan since I'm not a ninja of tf.device.

### PR DESCRIPTION
**Pull request by gabrieldemarmiesse**
Needs #1767 to be merged first this will make the diff smaller.

Closes #1682 

tf.test.is_gpu_available() is deprecated and was causing issues so I replaced it.

Maybe we should replace the name `device` for the fixture? Maybe it's strange that there is a device being a strategy.

I think we'll rename `with_device` -> `with_scope` and `device` -> `scope`. But the renaming would make a big diff, so it's a TODO for a next pull request.

### CODE REVIEW

**Comment by Squadrick** [comment]
@gabrieldemarmiesse What's blocking this PR?

**Comment by gabrieldemarmiesse** [comment]
#1767

**Comment by guillaumekln** [comment]
On this subject, I've been using this trick to test distribution strategies locally:

https://github.com/OpenNMT/OpenNMT-tf/blob/v2.10.0/opennmt/tests/test_util.py#L41-L61

It's a decorator that resets the TensorFlow context and creates 2 logical CPU devices for the test being run.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -445,6 +447,24 @@ def test_something2(device):
...
+
+@pytest.mark.with_device(["cpu", "gpu", tf.distribute.MirroredStrategy])
+def test_something3(device):
+    # the code here will run three times, once on gpu, once on cpu and once with 
+    # a mirror distributed strategy.
+    # device will be "cpu:0" or "gpu:0" or "gpu:1" or "gpu:2" or the strategy.
+    # with the MirroredStrategy, it's equivalent to:
+    # strategy = tf.distribute.MirroredStrategy(...)
+    # with strategy.scope():
+    #     test_function(strategy)
```
Comment (seanpmorgan) [inline CONTRIBUTING.md:459]:
```suggestion
    # the code here will run three times, once on gpu, once on cpu and once with 
    # a mirrored distribution strategy on two virtual gpu devices.
    # `device` will be either "cpu:0", "gpu:0", or  "gpu:0" / "gpu:1" respectively.
    # With the MirroredStrategy, it's equivalent to:
    # strategy = tf.distribute.MirroredStrategy(...)
    # with strategy.scope():
    #     test_function(strategy)
```
Slight re-wording. Per the `gpus_for_testing` function I'm not sure how we'd get a gpu:1 and gpu:2:
```
RMSprop/RMSprop/update_1/update_1/sub/x: (Const): /job:localhost/replica:0/task:0/device:GPU:1
RMSprop/RMSprop/update_0/Const: (Const): /job:localhost/replica:0/task:0/device:GPU:0
```

Comment (gabrieldemarmiesse) [reply CONTRIBUTING.md:459]: Yeah, actually device is going to be cpu:0 or gpu:0 or the strategy. Let me reword that. Device can be the strategy too.

**Pull request merged**